### PR TITLE
http: refactory Http::Utility::resolveMostSpecificPerFilterConfig method

### LINKED ISSUE #14894: Inconsistency in per-route filter config resolution across filters
Some filters use `resolveMostSpecificPerFilterConfig()`, others use `mostSpecificPerFilterConfigTyped`. They both have independent logic around most-specific route config resolution, which is hard to follow and possibly keep inline. Is it plausible we can converge these?

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: http: refactory `Http::Utility::resolveMostSpecificPerFilterConfig` method
Additional Description:
The `Http::Utility::resolveMostSpecificPerFilterConfig` and `RouteEntry::mostSpecificPerFilterConfigTyped` are doing
the same thing. `RouteEntry::mostSpecificPerFilterConfigTyped` is more efficient, and `Http::Utility::resolveMostSpecificPerFilterConfig` can take care some of sanity checks. So converge them. Refactory `Http::Utility::resolveMostSpecificPerFilterConfig` to use
`RouteEntry::mostSpecificPerFilterConfigTyped`. And using `Http::Utility::resolveMostSpecificPerFilterConfig` for the filters consistently.
Risk Level: low
Testing: unittest
Docs Changes: n/a
Release Notes: n/a
Fixes #14894

### CODE REVIEW

**Review by alyssawilk — COMMENTED** [review]
Nice find!  Want to ping back when CI is happy and I'll take a look?

**Comment by soulxu** [comment]
> Nice find! Want to ping back when CI is happy and I'll take a look?

thanks for having a look at, let me make the CI happy first!

**Review by alyssawilk — APPROVED** [review]

**Comment by alyssawilk** [comment]
looks like this needs a main merge though!

**Review by htuch — COMMENTED** [review]
Makes sense to me; this is less confusing. Will defer to @alyssawilk for full review. Thanks!

**Comment by soulxu** [comment]
emm...coverage test failed. probably due to I removed few lines code.

**Review by alyssawilk — APPROVED** [review]
